Add Documentation for Elaborator Reflection

martinbaker opened the pull request:
This adds new pages to the documentation to attempt to explain Elaborator Reflection.

Existing documentation pages are not changed except:

docs/index.rst - links are added to the new pages and
the section title is changed to 'Theorem Proving and Elaborator Reflection' to reflect the new content.

docs/conf.py - added table width fix for Read the Docs Sphinx theme as explained here:
https://rackerlabs.github.io/docs-rackspace/tools/rtd-tables.html
Without this change the tables would be extremely wide which would make them unusable.

david-christiansen commented:
Thanks for doing this! I'll review soon!

martinbaker commented:
@david-christiansen

Thank you for your comments so far.

I'll work through and resolve your detailed comments, at first glance, I tend to agree with most of them.

Regarding the high level-concerns:

I agree with the general principle that content should not be duplicated but I also think its a bad principle that new users should be forced to read the source code (or API doc) to find what commands are available.
In the case of:
https://github.com/martinbaker/Idris-dev/blob/mydoc/docs/proofs/primative.rst
The problem I have (and I'm assuming other people may have) is to get an overview of all the operators (tactics) so I really do want to have all the tactics together, in one big table, so I stand the best chance of choosing the correct one. If a (meta) programmer already has some clue about what the need then they can use the search tools but new users don't know what to search for. Even for experienced users, I think its valuable to have all the information together in a compact form.

As you say "a lot of this text is copied and pasted from the built-in documentation, and then modified and extended" but I would like this to only be a first step to documentation aimed at users rather than developers. I think there is a lot more that needs to be done here but I this is the best I can do for now. What I would really like is a diagram for each tactic showing the state before, the state after and arrows between them so users can pick out what they need visually.

Regarding pictures:

I can include a SVG and a PNG file for each picture. The SVG file, being vector graphics, is best for drawing and modifying (using Inkscape) and PNG, being raster graphics, gives more consistent rendering. I can add a readme explaining how to generate the PNG from the SVG.

I appreciate that everyone is different in what type of documentation appeals to them. Personally I'm put off by big blocks of text but I find visual clues really help me find what I'm looking for. I also imagine that people would be using the documentation in different ways, I imagine a high proportion would be quickly scanning to try to find a particular piece of information.

zenntenn commented:
In terms of accessibility of images, including images is fine, as long as they are tagged properly so those using screen readers understand what they are, and that someone using a screen reader can still get all the information they need from the documentation without them. Basically, a visual aid to help grasp the text better but contributes no new information is fine, at least when it comes to accessible best practices.

nightscape commented on docs/proofs/elabReflection.rst:
```diff
@@ -0,0 +1,55 @@
+
+Idris provides a mechanism to modify the language without having to recompile Idris itself. We can think of this in terms of metaprogramming or domain specific languages or just building in new capabilities.
+
+In order to extend the language we need to know something about how Idris is complied. This page explains only what is needed to customise the elaboration. For more information about the compiler's implementation see `Edwin Brady's 2013 paper`_ and for customising the elaboration process see `Elaborator reflection: extending Idris in Idris`_ and `David Christiansen's PhD thesis`_ .
```
`complied` => `compiled`

martinbaker replied:
changed as suggested.

martinbaker commented on docs/proofs/elabReflection.rst:
```diff
@@ -0,0 +1,55 @@
+
+Idris provides a mechanism to modify the language without having to recompile Idris itself. We can think of this in terms of metaprogramming or domain specific languages or just building in new capabilities.
+
+In order to extend the language we need to know something about how Idris is complied. On this page we try to limit this explanation to only what we need to customise the elaboration. For more details of the compiler see `Edwin Brady 2013 paper`_ and for customising the elaboration process see `David Christiansen Practical Reflection and Metaprogramming`_ and  `Elaborator reflection: extending Id ...
```
Yes, I overused 'we' intending it to mean 'author and reader'. I think I have corrected that now.

martinbaker commented on docs/proofs/elabReflection.rst:
```diff
@@ -0,0 +1,55 @@
+
+Idris provides a mechanism to modify the language without having to recompile Idris itself. We can think of this in terms of metaprogramming or domain specific languages or just building in new capabilities.
+
+In order to extend the language we need to know something about how Idris is complied. On this page we try to limit this explanation to only what we need to customise the elaboration. For more details of the compiler see `Edwin Brady 2013 paper`_ and for customising the elaboration process see `David Christiansen Practical Reflection and Metaprogramming`_ and  `Elaborator reflection: extending Id ...
```
changed as suggested.

martinbaker commented on docs/proofs/elabReflection.rst:
```diff
@@ -0,0 +1,55 @@
+
+Idris provides a mechanism to modify the language without having to recompile Idris itself. We can think of this in terms of metaprogramming or domain specific languages or just building in new capabilities.
+
+In order to extend the language we need to know something about how Idris is complied. On this page we try to limit this explanation to only what we need to customise the elaboration. For more details of the compiler see `Edwin Brady 2013 paper`_ and for customising the elaboration process see `David Christiansen Practical Reflection and Metaprogramming`_ and  `Elaborator reflection: extending Id ...
```
changed as suggested.

martinbaker commented on docs/proofs/elabReflection.rst:
```diff
@@ -0,0 +1,55 @@
+
+Idris provides a mechanism to modify the language without having to recompile Idris itself. We can think of this in terms of metaprogramming or domain specific languages or just building in new capabilities.
+
+In order to extend the language we need to know something about how Idris is complied. On this page we try to limit this explanation to only what we need to customise the elaboration. For more details of the compiler see `Edwin Brady 2013 paper`_ and for customising the elaboration process see `David Christiansen Practical Reflection and Metaprogramming`_ and  `Elaborator reflection: extending Id ...
```
changed as suggested.

martinbaker commented on docs/proofs/elabReflection.rst:
```diff
@@ -0,0 +1,55 @@
+
+Idris provides a mechanism to modify the language without having to recompile Idris itself. We can think of this in terms of metaprogramming or domain specific languages or just building in new capabilities.
+
+In order to extend the language we need to know something about how Idris is complied. On this page we try to limit this explanation to only what we need to customise the elaboration. For more details of the compiler see `Edwin Brady 2013 paper`_ and for customising the elaboration process see `David Christiansen Practical Reflection and Metaprogramming`_ and  `Elaborator reflection: extending Id ...
```
moved to tactics.rst

martinbaker replied:
moved back to elabReflection.rst, but to later in the page, where holes and guesses are discussed.

martinbaker commented on docs/proofs/primative.rst:
```diff
@@ -0,0 +1,517 @@
+         lookupFunDefnExact n = case !(lookupFunDefn n) of
+           [res] => pure res
+           []    => fail [TextPart
+               "No function named", NamePart n]
```
It is a code block inside a table and so there a lot of indents. I don't think it matters because the reader will see the formatted version.

martinbaker replied:
I have now removed this code

martinbaker commented on docs/proofs/primative.rst:
```diff
@@ -0,0 +1,517 @@
+
+       .. code-block:: idris
+
+         lookupFunDefnExact n = case !(lookupFunDefn n) of
```
removed from file.

martinbaker commented on docs/proofs/primative.rst:
```diff
@@ -0,0 +1,517 @@
+
+       tryCatch : Elab a -> (Err -> Elab a) -> Elab a
+
+       Fixme - following does not work. 
```
I was just trying to get a small, self contained, example of 'tryCatch' just so the user can see how to use it.
Eventually I would like to have a small example, like this, for every tactic.

martinbaker replied:
code fixed and FIXME removed.

martinbaker commented on docs/proofs/elabReflection.rst:
```diff
@@ -0,0 +1,55 @@
+
+Here is a diagram of the multistage process at the top level when Idris code gets compiled:
+
+.. image:: ../image/idrisTopLevel.png
```
I've changed some of the arrows, hopefully that's better?

I have also uploaded the SVG source files with instructions for editing.

martinbaker replied:
Regarding the antialiasing effect: I am not seeing that on the editors and viewers that I am using. Most importantly I am not seeing it in the html generated by sphinx, which I suspect most readers will use, so I will try to get the lines straight to avoid distracting the reader with things not being straight but I don't think it is a big problem.

martinbaker commented on docs/proofs/elabReflection.rst:
```diff
@@ -0,0 +1,55 @@
+
+In order to extend the language we need to know something about how Idris is complied. On this page we try to limit this explanation to only what we need to customise the elaboration. For more details of the compiler see `Edwin Brady 2013 paper`_ and for customising the elaboration process see `David Christiansen Practical Reflection and Metaprogramming`_ and  `Elaborator reflection: extending Id ...
+
+Here is a diagram of the multistage process at the top level when Idris code gets compiled:
```
changed as suggested.

martinbaker commented on docs/proofs/elabReflection.rst:
```diff
@@ -0,0 +1,55 @@
+
+TT is a core language which is syntactically very simple, this makes it easy for computers to process but very verbose and hard for humans to read. This elaboration is done by a logic language (proof tactics) similar to LTac in Coq. Here we use the word 'tactics' to refer to these elaboration tactics - not to be confused with the old tactics mechanism.
+
+.. image:: ../image/compareToProofAssist.png
```
I've made some changes to address this.

martinbaker commented on docs/proofs/elabReflection.rst:
```diff
@@ -0,0 +1,55 @@
+   :width: 484px
+   :height: 147px
+
+TT is a core language which is syntactically very simple, this makes it easy for computers to process but very verbose and hard for humans to read. This elaboration is done by a logic language (proof tactics) similar to LTac in Coq. Here we use the word 'tactics' to refer to these elaboration tactics - not to be confused with the old tactics mechanism.
```
done

martinbaker commented on docs/proofs/elabReflection.rst:
```diff
@@ -0,0 +1,55 @@
+   :width: 484px
+   :height: 147px
+
+TT is a core language which is syntactically very simple, this makes it easy for computers to process but very verbose and hard for humans to read. This elaboration is done by a logic language (proof tactics) similar to LTac in Coq. Here we use the word 'tactics' to refer to these elaboration tactics - not to be confused with the old tactics mechanism.
```
I have made these changes, slightly reluctantly, because I am concerned that a person reading the documentation will first come across the word 'tactics' to refer to the deprecated tactics code. I would like some way to make it clear that not what we are talking about when we use the word 'tactics' here.

martinbaker commented on docs/proofs/elabReflection.rst:
```diff
@@ -0,0 +1,55 @@
+   :width: 229px
+   :height: 114px
+
+This gives the opportunity to allow the primitives of the elaboration mechanism  to be exposed to metaprograms.
```
changed as suggested.

martinbaker commented on docs/proofs/elabReflection.rst:
```diff
@@ -0,0 +1,55 @@
+This gives the opportunity to allow the primitives of the elaboration mechanism  to be exposed to metaprograms.
+During elaboration TT (Raw) structure contains:
+
+- holes - terms yet to be inferred.
```
changed as suggested.

martinbaker commented on docs/proofs/elabReflection.rst:
```diff
@@ -0,0 +1,55 @@
+During elaboration TT (Raw) structure contains:
+
+- holes - terms yet to be inferred.
+- guesses - Suppresses β-reduction while precise structure not known.
```
changed as suggested.

martinbaker commented on docs/proofs/elabReflection.rst:
```diff
@@ -0,0 +1,55 @@
+   :width: 268px
+   :height: 219px
+
+As already mentioned the TT core language is kept syntactically very simple. Part of the reason for this is that its correctness is already well proven using logic. For instance, here are the binders in TT with corresponding code and logic type introduction rules:
```
removed offending sentence.

martinbaker commented on docs/proofs/elabReflection.rst:
```diff
@@ -0,0 +1,55 @@
+
+Type checker:
+
+- include universe levels
```
removed as suggested

martinbaker commented on docs/proofs/elabReflection.rst:
```diff
@@ -0,0 +1,55 @@
+- include universe levels
+- distinguish between global and local bound names.
+
+.. image:: ../image/elabOverview.png
```
I have updated this diagram and the text around it. The aim of the diagram is to give a top-level view of how the tactics relate back to the TT code generated, I think it could still be improved but I haven't yet worked out how. However I think its a lot better than it was.

martinbaker commented on docs/proofs/elabReflection.rst:
```diff
@@ -0,0 +1,55 @@
+
+.. list-table::
+
+   * - There is an elaborator for both definitions and terms, the definition elaborator calls the term elaborator when required.
```
Removed diagram for now.

martinbaker commented on docs/proofs/elabReflection.rst:
```diff
@@ -0,0 +1,55 @@
+
+As already mentioned the TT core language is kept syntactically very simple. Part of the reason for this is that its correctness is already well proven using logic. For instance, here are the binders in TT with corresponding code and logic type introduction rules:
+
+.. image:: ../image/binders.png
```
changed the text to reflect these concerns.

martinbaker commented on docs/proofs/index.rst:
```diff
@@ -1,10 +1,10 @@
-Theorem Proving
-###############
+#########################################
+Theorem Proving and Elaborator Reflection
```
I could easily change this if you want.
I'm not sure myself. It seems to me that there is a lot of things (like tactics) that are common to both. Perhaps tactics should be separated out into a third section?
I think that, in the future, a lot more needs to be added to this so perhaps it justifies more sections. In fact I think someone should write a whole book on this subject!

martinbaker commented on docs/proofs/example1.rst:
```diff
@@ -0,0 +1,280 @@
+Example 1
```
I changed the title to 'Elaborator Reflection - Identity Example' because I think that, at some stage in the future, there should be other examples added.

martinbaker commented on docs/proofs/example1.rst:
```diff
@@ -0,0 +1,280 @@
+Example 1
+=========
+
+.. list-table:: Nat identity
```
I removed 'Nat identity' because I don't think the table really needs a title.

martinbaker commented on docs/proofs/example1.rst:
```diff
@@ -0,0 +1,280 @@
+
+.. list-table:: Nat identity
+
+   * - As an example of elaborator reflection here we step through this identity implementation:
```
I'm finding it quite hard to replace 'we' because 'I' is not appropriate due to the anonymous nature of the documentation so I am rewording sentences in a more impersonal way.

martinbaker commented on docs/proofs/example1.rst:
```diff
@@ -0,0 +1,280 @@
+         %language ElabReflection
+
+         idNat : Nat -> Nat
+         idNat = %runElab (do
```
OK, I've done that here. I think there might be some cases where, for space reasons, I might have to slightly reduce the amount of indent.

martinbaker commented on docs/proofs/example1.rst:
```diff
@@ -0,0 +1,280 @@
+
+.. list-table:: Nat identity
+
+   * - We start off with a hole for the whole term (of type Nat->Nat).
```
changed

martinbaker commented on docs/proofs/primative.rst:
```diff
@@ -0,0 +1,517 @@
+Primitive Operators
```
file renamed

martinbaker replied:
renamed file

martinbaker commented on docs/proofs/primative.rst:
```diff
@@ -0,0 +1,517 @@
+
+       debug : Elab a
+
+       Only seems to compile if 'debug' is last tactic so comment out following tactics like this: 
```
corrected as suggested.

martinbaker commented on docs/proofs/primative.rst:
```diff
@@ -0,0 +1,517 @@
+
+       debugMessage : (msg : List ErrorReportPart) -> Elab a 
+
+       Only seems to compile if 'debug' is last tactic so comment out following tactics like this: 
```
corrected as suggested.

martinbaker commented on docs/proofs/primative.rst:
```diff
@@ -0,0 +1,517 @@
+
+       fail : List ErrorReportPart -> Elab a
+
+       Return type is given by last term so fail cant be at the the last term, we would expect fail to be in some sort of conditional statement:
```
reworded to make sense.

martinbaker commented on docs/proofs/example1.rst:
```diff
@@ -0,0 +1,280 @@
+   * - Introduce a lambda binding around the current hole and focus on the body.
+     - intro \`{{x}}
+
+   * - state now looks like this
```
changed.

martinbaker commented on docs/proofs/example1.rst:
```diff
@@ -0,0 +1,280 @@
+.. list-table:: Nat identity
+
+   * - We start off with a hole for the whole term (of type Nat->Nat).
+       We fill that with 'intro' which creates a lambda term with a hole for the  expression.
```
added this text

martinbaker commented on docs/proofs/example1.rst:
```diff
@@ -0,0 +1,280 @@
+       We fill that with 'intro' which creates a lambda term with a hole for the  expression.
+     - diagram
+
+       .. image:: ../image/tree.png
```
Yes, its not a very good diagram, I'll think about it more.

martinbaker commented:
@david-christiansen

In one of your comments for this pull request:
https://github.com/idris-lang/Idris-dev/pull/4700#discussion_r284101057
you commented on the strange syntax that I used to display the results of getEnv, getGoal, getHoles and getGuess in the walkthrough example.

Well, I wanted a way to see the exact contents and syntax of these results, I could not find a way to do this apart from writing this horrible hack here:

https://github.com/martinbaker/Idris-dev/blob/uglyTTPrinter/libs/prelude/Language/Reflection/TTPrinter.idr

What is the recommended way to see and display the full results from getEnv, getGoal, getHoles and getGuess? I am looking for something that can be used from within Idris (so it can be displayed by debugMessage or fail) not just for ElabReflection but also to help me get a more intuitive understanding of the syntax of TT and some of the internal workings.

I see you have worked on pretty printers so I'm sure there must be something that does this.

Thanks, Martin

martinbaker commented on docs/proofs/generatingData.rst:
```diff
@@ -0,0 +1,308 @@
+  %language ElabReflection
+
+  addData : Elab ()
+  addData = do
```
I have changed most occurrences of this but there is one case remaining where I cant indent that far to the right due to space issues.

martinbaker commented on docs/proofs/generatingData.rst:
```diff
@@ -0,0 +1,308 @@
+        Constructor `{{MkN}} [MkFunArg `{{x}} `(Nat) Implicit NotErased]
+            (RApp (Var `{{DataDef.N}}) (Var `{{x}})),
+        Constructor `{{MkN'}} [MkFunArg `{{x}} `(Nat) Explicit NotErased]
+            (RApp (Var `{{DataDef.N}}) (RApp (Var `{S}) (Var `{{x}})))
```
changed

martinbaker commented on docs/proofs/generatingData.rst:
```diff
@@ -0,0 +1,308 @@
+Generating Data and Functions at Compile Time
+=============================================
+
+We can  construct a data structure at compile-time in an Elab monad.
```
changed

martinbaker commented on docs/proofs/generatingData.rst:
```diff
@@ -0,0 +1,308 @@
+  λΠ> :t MkN
+  MkN : N x
+
+Here is another, more complicated, example from the same test file. This also creates a function.
```
changed the examples

martinbaker commented on docs/proofs/generatingData.rst:
```diff
@@ -0,0 +1,308 @@
+   * - addImplementation
+     - Register a new implementation for interface resolution.
+
+       - @ ifaceName the name of the interface for which an implementation is being registered
```
I was hoping to use '@' as a shorthand to itemise the arguments but I've changed it now.

martinbaker commented on docs/proofs/example1.rst:
```diff
@@ -0,0 +1,280 @@
+
+         idNat : Nat -> Nat
+         idNat = %runElab (do
+           intro `{{x}}
```
quasiquotation is explained in: docs/proofs/tactics.rst

martinbaker commented on docs/proofs/generatingData.rst:
```diff
@@ -0,0 +1,308 @@
+This can allow proofs to be generated for user defined types or it could allow types to be automatically generated to support user defined types.
+An example is the code, from `Elaborator reflection: extending Idris in Idris`_, that automatically generates accessibility predicates using the Bove-Capretta method.
+The following simple outline example is adapted from
+`<https://github.com/idris-lang/Idris-dev/blob/master/test/meta002/DataDef.idr>`_
```
removed this

martinbaker commented on docs/proofs/generatingData.rst:
```diff
@@ -0,0 +1,308 @@
+         λΠ> el Base
+         Bool : Type
+
+So these are the functions that we can use to create data and functions in the Elab monad:
```
At the moment there is some duplication but think the explanations need to be greatly expanded for the documentation and I think this gives a framework to do it in manageable stages.

martinbaker commented on docs/proofs/primative.rst:
```diff
@@ -0,0 +1,517 @@
+Primitive Operators
+===================
+
+.. list-table:: Primitive Operators
```
I think the API document falls far short of the information required to metaprogram using this code. I think the documentation also needs examples, diagrams and such that can't go in the API. I would therefore like to keep this table as a framework for future enhancements to the documentation.

martinbaker commented on docs/proofs/example1.rst:
```diff
@@ -0,0 +1,280 @@
+         idNat : Nat -> Nat
+         idNat = %runElab (do
+
+   * - In order to investigate how the program works we can look at the proofState at each stage as the tactics are applied. So here is the proofState at the start:
```
I have put the explanation of the terminology issue in the tactics.rst page.

martinbaker commented on docs/proofs/example1.rst:
```diff
@@ -0,0 +1,280 @@
+   * - In order to investigate how the program works we can look at the proofState at each stage as the tactics are applied. So here is the proofState at the start:
+     - state
+
+       .. image:: ../image/elabProofStateEx1_1.png
```
state is described in tactics.rst which comes before this page.

martinbaker commented on docs/proofs/example1.rst:
```diff
@@ -0,0 +1,280 @@
+          :width: 310px
+          :height: 124px
+
+   * - We can also look at the logic at each stage: 
```
The aim is to illustrate the tactics by analogy with proofs, as a line with the premises above it and the conclusion below it. The conclusion is the hole we are trying to fill and the premises may be sub-goals.

martinbaker commented on docs/proofs/example1.rst:
```diff
@@ -0,0 +1,280 @@
+          :width: 126px
+          :height: 33px
+
+   * - The term is:
```
Well, I wanted a way to see the exact contents and syntax of these results, I could not find a way to do this apart from writing this horrible hack here:

https://github.com/martinbaker/Idris-dev/blob/uglyTTPrinter/libs/prelude/Language/Reflection/TTPrinter.idr

What is the recommended way to see and display the full results from getEnv, getGoal, getHoles and getGuess? I am looking for something that can be used from within Idris (so it can be displayed by debugMessage or fail) not just for ElabReflection but also to help me get a more intuitive understanding of the syntax of TT and some of the internal workings.

martinbaker commented on docs/proofs/example1.rst:
```diff
@@ -0,0 +1,280 @@
+   * - state now looks like this
+     - state
+
+       .. image:: ../image/elabProofStateEx1_2.png
```
I think a picture is more likely to help intuitive understanding. I would really like to see env, goal, holes and guess grouped together in the state.
I have changed the diagram so the lines all go somewhere.

martinbaker commented on docs/proofs/generatingData.rst:
```diff
@@ -0,0 +1,308 @@
+
+.. code-block:: idris
+
+  data N : Nat -> Type where
```
done

martinbaker commented on docs/proofs/generatingData.rst:
```diff
@@ -0,0 +1,308 @@
+      Constructor `{{Base}} [] (Var U),
+      Constructor `{{Pi}}
+        [MkFunArg `{{code}} (Var U) Explicit NotErased,
+          MkFunArg `{{body}} `(~(RApp (Var el) (Var `{{code}})) -> ~(Var U)) Explicit NotErased]
```
Its described on the tactics.rst page

martinbaker commented on docs/proofs/example1.rst:
```diff
@@ -0,0 +1,280 @@
+          :height: 33px
+
+   * - The term is:
+     - ?{hole_0} ≈ ? {hole_2} . {hole_2} . {hole_0}
```
it is discussed on the tactics.rst page.

martinbaker commented:
I think I have made the changes to this pull request that I intend to do, for now.

It is getting very unwieldy and it would be good to get it to where you could accept it. Then, future updates could be done in more digestible chunks.

I think its a lot better than it was due to your helpful reviews.

I think a lot more work needs to be done, to this documentation, before it is in a state where it presents all the information a potential meta-programmer needs. However, any contribution I can make would be easier if spread out into smaller sub-goals.

Martin

david-christiansen commented:
I'll look again as soon as I get some free time. Thanks for the hard work!

Den tor. 23. maj 2019 kl. 04.28 skrev martinbaker <notifications@github.com
>:

> I think I have made the changes to this pull request that I intend to do,
> for now.
>
> It is getting very unwieldy and it would be good to get it to where you
> could accept it. Then, future updates could be done in more digestible
> chunks.
>
> I think its a lot better than it was due to your helpful reviews.
>
> I think a lot more work needs to be done, to this documentation, before it
> is in a state where it presents all the information a potential
> meta-programmer needs. However, any contribution I can make would be easier
> if spread out into smaller sub-goals.
>
> Martin
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/idris-lang/Idris-dev/pull/4700?email_source=notifications&email_token=AAA4FASZKSPG3MO65BT35P3PWZ5XVA5CNFSM4HLTFEBKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODWB5OVY#issuecomment-495179607>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAA4FAST3TFY2R4HMZKOHT3PWZ5XVANCNFSM4HLTFEBA>
> .
>

martinbaker commented:
I have now updated the diagrams, in the example, which show the holes and what they depend on.

They now separate the holes into different columns, I think this makes it much clearer.
I have also removed the confusing references to 'logic diagrams'.

Hopefully this documentation now explains elaborator reflection better.

david-christiansen commented:
Thanks for keeping at this! I haven't forgotten you - I've just had a lot of work and travel. But the class I was teaching has finished, and my trip is done, so I should have some time this coming weekend.

Sorry for the delay.

martinbaker commented:
> so I should have some time this coming weekend.

Thats great!

I have just added a page about the notation for holes and guesses, I did not want the helpful information you gave to my forum question to get lost.

Regarding the elaborator reflection documentation as a whole: I am hoping that this pull request could just be the start, I think that a lot more documentation is needed, for example:

* More information is needed describing individual tactics such as 'attack'. The sort of information that is in your joint paper with Edwin. In addition I would really like to see a minimal example for each tactic. I would also like to have a diagram showing showing how each tactic modifies the state, I am just trying to find a fast way for a user to find the tactic required without searching lots of wordy text. I know you are concerned that what I have done, so far, duplicates the API documentation with some additional information. I just think a potential user needs a lot more than is in the API and what I have done so far is just intended as a framework to be improved later.

* I think a more motivational introduction is required to elaborator reflection and metaprogramming, when it is useful to customise the language, various methods of implementing domain specific languages and so on. I think this is best coming from you (but I could try to paraphrase your thesis if you want!).

I know you would prefer this all to be in a top-level metaprogramming section of its own, separated from proofs. But where would common information (like tactics) go? Should tactics should be separated out into a third section?
I think a rearrangement will be required, at the moment there is:
* A page about theorem proving in the introduction section.
* A top-level section about theorem proving and elaborator reflection.
* A page about elaborator reflection in the language reference section.
So a lot of swapping around does seem to be required but would it be simpler and less messy to do that as a separate exercise later?

By the way, I have been checking the html produced by running sphinx from the command line like this:

sphinx-build -b html Idris-dev/docs IdrisDoc

You may find there are less artifacts in the diagrams when viewed this way?

I may be using a newer version of sphinx than readthedocs, but I have used very standard markup, so it seems unlikely to make a difference.

Martin

melted commented:
I'm pretty sure the travis failure is because this is based on a commit where stack wasn't working. So it should be fine.

martinbaker commented:
I was hoping to extend and improve this documentation further but I don't really want to do more until I know if this pull request will be accepted.

I guess you are busy on other things but I'm not sure how to allocate my own time at the moment. If this pull request is not helpful please let me know as I don't want to waste your time and mine.

Martin

melted commented:
No worries about it being accepted, it's going in. But yeah, let's not put more stuff into the PR now, so we can merge something. It's always possible to expand it later.

@david-christiansen Any time for review?

david-christiansen commented:
I hope to have time soon! I've been feeling quite overwhelmed with everything lately - sorry!

melted commented:
This PR represents too much work to be left hanging, so I'll merge it.
